keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +337,302 @@ or later. The dual-stack feature requires no changes to this API.
+KUBERNETES_IPv4_PORT_443_TCP_PROTO=tcp
+KUBERNETES_IPv4_SERVICE_HOST<IPv6>
+KUBERNETES_IPv4_SERVICE_PORT=443
+KUBERNETES_IPv4_SERVICE_PORT_HTTPS=443
+``` 
 
 #### Type NodePort, LoadBalancer, ClusterIP
-Services will only be assigned an address from a single address family (either IPv4 or IPv6). The address family for the Service’s cluster IP is determined by setting the field, `.spec.ipFamily`, on that Service. You can only set this field when creating a new Service. Setting the `.spec.ipFamily` field is optional and should only be used if you plan to enable IPv4 and IPv6 Services and Ingresses on your cluster. The configuration of this field not a requirement for egress traffic. The default address family for your cluster is the address family of the first service cluster IP range configured via the `--service-cluster-ip-range` flag to the kube-controller-manager. The `.spec.ipFamily` field can be set to either IPv4 or IPv6.
+In a dualstack cluster Services can be:
+1. Single Stack: service ip are allocated (or reserved, if provided by user) from the first or the second cidr (if configured) entry as configured using `--service-cluster-ip-range` flag, in this scenario services will have a single allocated `ClusterIP` (also set in `ClusterIPs`).
+2. Dual Stack (optional): service ip is allocated (or reserved, if provided by user) from both the first and the second cidr (if configured) entries. In this scenario services will have a dual assigned `ClusterIPs`. If the cluster is not configured for dualstack then  the resulting service will be created as a single stack and `ClusterIP/ClusterIPs` will be assigned from a single family.
+3. Dual Stack (required): if the cluster is not configured for dual stack then service creation will fail. If the cluster is configured for dual stack the resulting service will carry two dual stack `ClusterIPs`.
+
+The above is achieved using the following changes to Service api.
+1. Add a `spec.preferDualStack` optional boolean field.
+2. Add a `spec.ipFamilies[]` optional field with possible values of `IPv4`, `IPv6` for each each entry or `nil` for the entire slice. Maximum entries in this array is two and each entry must be `IPv4`or `IPv6` in any order. This field identifies the expected ip families of a service. api-server will reject the service creation request if the ip family is not supported by the cluster. Further dependency and validation on this field is described below.
+3. Add a `spec.ClusterIPs[]` field that follows the same rules as `status.PodIPs/status.PodIP`. Maximum entries in this array is two entries of two different ip address families. Semantically `spec.ClusterIP` will be referred as `Primary Service IP`.
+
+> Services are expected to be semantically valid whenever values for the optional fields are provided. An example a service with `spec.ipFamilies : [ ipv4 ]` and `spec.CluterIPs: [ 'some-ipv6' ]` is considered invalid because the family does not match the `ClusterIPs[0]`.
```

Comment:
typo "CluterIPs". (also elsewhere)

maybe provide an actual IPv6 address there?

actually `[ ipv4 ]` is wrong too it should be `[ "IPv4" ]`